8286058: AArch64: clarify types of calls

eastig opened the pull request:
The PR clarifies the types of calls AArch64 OpenJDK uses. It cleans up far_call, far_jump and trampoline_call. It removes trampoline_call1 because its use cases are now supported by trampoline_call.
Tested a fastdebug build:
- `gtest`: Passed
- `tier1`...`tier3`: Passed

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8286058](https://bugs.openjdk.java.net/browse/JDK-8286058): AArch64: clarify types of calls


### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8564/head:pull/8564` \
`$ git checkout pull/8564`

Update a local copy of the PR: \
`$ git checkout pull/8564` \
`$ git pull https://git.openjdk.java.net/jdk pull/8564/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8564`

View PR using the GUI difftool: \
`$ git pr show -t 8564`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8564.diff">https://git.openjdk.java.net/jdk/pull/8564.diff</a>

</details>

eastig commented:
@vnkozlov, @theRealAph
Could you please review?

vnkozlov reviewed (approved):
Good.

vnkozlov commented:
I will run tests too to make sure we don't hit asserts.

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:
```diff
@@ -1084,6 +1084,34 @@ class MacroAssembler: public Assembler {
+  // with the call of the target.
+  //
+  // Trampoline_call is most suitable for calls of Java methods. Java calls callees can be changed
+  // to the interpreter or different versions of a compiled method. Those callees can be
```
This is not true. Trampoline calls are needed for calls to the runtime too.

eastig replied:
Why? What are the cases? This is not clear from sources.

eastig replied:
> This commentary is garbled, misleading, and very confusing.

Could you please be more neutral in your words?

> I suspect that anyone confused about how trampoline calls work will be even more confused after trying to read this. 

This is why we have the review process to make things better.

> Can you please explain what aspects of trampolines you're trying to clarify? Let's have a conversation about how to document this.

I am trying to clarify why we have two ways for long calls, pros/cons, which of two can be optimized. This mechanism of long call is very important part but it is not documented.

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:
```diff
@@ -1084,6 +1084,34 @@ class MacroAssembler: public Assembler {
+  // The code for runtime calls can also be generated with far_call. For possible far-distant callees
+  // far_call does not use the stub code section for additional code. It inserts the code at a call site.
+  // This prevents the call from optimization to a direct call when the code is copied to CodeCache.
+  //
```
I can't understand any of this. As far as I can tell it is nonsense.
It introduces pointless confusing terminology such as "far-distant" and "near-distant".
"This prevents the call from optimization to a direct cal" is completely wrongl

eastig replied:
> "This prevents the call from optimization to a direct cal" is completely wrongl

If I am wrong, could you please point me at the place where 'adrp, add, bl' is optimized?

> As far as I can tell it is nonsense.

Could you please be more constructive?

> It introduces pointless confusing terminology such as "far-distant" and "near-distant".

Please, offer your variant. There is no standard terminology:
- GCC uses: long/short https://gcc.gnu.org/onlinedocs/gcc-8.1.0/gcc/ARM-Function-Attributes.html
- TI uses: relative/absolute/far away/near call https://downloads.ti.com/docs/esd/SPNU118/generate-far-call-trampolines-trampolines-option-stdz0755016.html
- ARM always calls tramopolines veneers: https://developer.arm.com/documentation/dui0803/d/pge1406301797482

We can use: short/long like GCC. What do you think?

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:
```diff
@@ -1084,6 +1084,34 @@ class MacroAssembler: public Assembler {
+  //
+  // If a mark of the generated call BL is needed, a pointer to CodeBuffer keeping the generated code
+  // must be provided.
+  //
```
What does "If a mark of the generated call BL is needed" mean?

eastig replied:
This is what is done in the function: 
```
address MacroAssembler::trampoline_call(Address entry, CodeBuffer* cbuf = NULL) {
...
  if (cbuf) cbuf->set_insts_mark();
  relocate(entry.rspec());
  if (!far_branches()) {
    bl(entry.target());
  } else {
    bl(pc());
  }
```

And most of the time NULL is passed.
I don't know how this code  passed review. It smells badly.
What written is based on what I see: how the function is used. For example, can `cbuf` be any `CodeBuffer`? If not, how is it connected to the current CodeBuffer? If it is the same, why we pass a pointer but not a flag.

theRealAph replied:
Sorry, I didn't realize what "mark of the generated call" meant.

Good point. I think it is likely that the `inst_mark` is no longer needed by the AArch64 back end. When we were experimenting with trampoline calls we did a lot of experiments, and I believe that `inst_mark` was needed at one time. However, I do not think that any current user of `trampoline_call()` uses the insts_mark for anything, and it could usefully be removed as part of a cleanup.

theRealAph reviewed (commented):
This commentary is garbled, misleading, and very confusing. I suspect that anyone confused about how trampoline calls work will be even more confused after trying to read this.
Can you please explain what aspects of trampolines you're trying to clarify? Let's have a conversation about how to document this.

eastig commented:
> This commentary is garbled, misleading, and very confusing.

Could you please be more neutral in your words?

> I suspect that anyone confused about how trampoline calls work will be even more confused after trying to read this. 

This is why we have the review process to make things better.

> Can you please explain what aspects of trampolines you're trying to clarify? Let's have a conversation about how to document this.

I am trying to clarify why we have two ways for long calls, pros/cons, which of two can be optimized. This mechanism of long call is very important part but it is not documented.

theRealAph commented:
> Could you please be more neutral in your words?

I apologize for my use of words. It was an inappropriate thing to say, and I wish I hadn't said it.

If you would like to continue, perhaps we could work on this together, and come up with something we both like. Would you be happy to try that?

I think it could look something like a list. Here's a rough sketch of what I think might work.

```

AArch64 OpenJDK uses four different types of call

direct call: bl address

  This is the shortest and the fastest, but only has a reach of ...

far call: adrp; add; bl reg

  This is longer and slower than a direct call, but can reach anywhere
  in the code cache...

trampoline call:

  This is only available in C2-compiled code. It is a combination of a
  direct call, which is used if the destination of a call is in range,
  and a register-indirect call. It has the advantages that it can
  reach anywhere in the AArch64 address space and that it is patchable
  at runtime when the generated code is being executed by other
  threads.

     bl trampoline

   Stub:

   trampoline:
     ldr reg, pc + 8
     br reg
     <64-bit destination address>   

indirect call: move reg, address; blr reg

   This too can reach anywhere in the address space, but it cannot be
   patch while code is running, so it must be used at a safepoint.

```

eastig commented:
Hi Andrew,

> If you would like to continue, perhaps we could work on this together, and come up with something we both like. Would you be happy to try that?

Sorry for the late response. I am on a business trip. Yes, I am keen to finish this.
BTW, we have a new `trampoline_call1`: https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp#L1101. This is a surprise from the project Loom. I'll try  to handle it as well.

> I apologize for my use of words. It was an inappropriate thing to say, and I wish I hadn't said it.

Thank you. Apologies accepted.

Thank you for the details. They helped a lot. Most of them are aligned with what I've read in sources and seen in a debugger.

> trampoline call:
...
  It has the advantages that it can
  reach anywhere in the AArch64 address space.

I created a bug: https://bugs.openjdk.java.net/browse/JDK-8286314. With small CodeCache, trampolines are not created for out of range targets which are outside CodeCache. 

> Ah, I think I see. Are you saying that a far call is not
converted to a direct call when the code is moved into the code
cache, even if a direct call might reach its target? And that far
calls do not need a trampoline.

I wanted to say that trampoline calls support link-time optimization: replacing a trampoline call by a direct one. Link-time optimization are not applied to far calls at the moment. Could I write in this way?

BTW, if we are not going to relocate code (this is true for non-nmethod), we can patch far calls as well. During copying code to CodeCache, we can:
- keep `adrp;add` but replace `blr` with `bl`
- replace `adrp;add;blr` with  `nop;nop;bl`

What do you think?

theRealAph commented:
Hi,

> Thank you for the details. They helped a lot. Most of them are aligned with what I've read in sources and seen in a debugger.
> 
> > trampoline call:
> > ...
> > It has the advantages that it can
> > reach anywhere in the AArch64 address space.
> 
> I created a bug: https://bugs.openjdk.java.net/browse/JDK-8286314. With small CodeCache, trampolines are not created for out of range targets which are outside CodeCache.

Oh! Thank you. I wonder why we never saw that one before. I guess because we don't normally have a small-enough CodeCache.

> I wanted to say that trampoline calls support link-time optimization: replacing a trampoline call by a direct one. Link-time optimization are not applied to far calls at the moment. Could I write in this way?

Sounds good.

> BTW, if we are not going to relocate code (this is true for non-nmethod), we can patch far calls as well. During copying code to CodeCache, we can:
> 
>     * keep `adrp;add` but replace `blr` with `bl`
> 
>     * replace `adrp;add;blr` with  `nop;nop;bl`
> 
> What do you think?

Either sounds fine. I guess the latter will be a bit more efficient.

eastig commented:
Hi Andrew,

JFYI

> However, I do not think that any current user of trampoline_call() uses the insts_mark for anything, and it could usefully be removed as part of a cleanup.

I have removed this. There were assert crashes in fastdebug because trampoline_call must be connected with CompiledStaticCall::emit_to_interp_stub to have correct relocInfo records. I fixed this.

> BTW, we have a new trampoline_call1: https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp#L1101. This is a surprise from the project Loom. I'll try to handle it as well.

I am currently removing `trampoline_call1`.

eastig commented:
Hi Andrew,

The final version is here. Could you please review it?

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -613,23 +609,31 @@ void MacroAssembler::call_VM_helper(Register oop_result, address entry_point, in
       // code during its branch shortening phase.
       CompileTask* task = ciEnv::current()->task();
+
+      // We need to be at phase output when we check the size of emitted code.
```
```suggestion
      // We need to be in the output phase when we check the size of emitted code.
```

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:
```diff
@@ -1113,10 +1113,51 @@ class MacroAssembler: public Assembler {
+  //     +/-128MB for the release build, +/-2MB for the debug build.
+  //
+  //   - far call: adrp reg, pc_relative_offset; add; bl reg
+  //     This is longer and slower than a direct call. The offset has
```
```suggestion
  //     This is longer than a direct call. The offset has
```
I wouldn't be so definite as to say it's slower. Let's stay with longer.

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:
```diff
@@ -1113,10 +1113,51 @@ class MacroAssembler: public Assembler {
+  //       br reg
+  //       <64-bit destination address>
+  //
+  //     A link-time optimization can be applied to a trampoline call when the generated
+  //     code is moved to the code cache. A trampoline call is replaced with by a direct call.
```
```suggestion
  //     If the destination is in range when the generated code is moved to the code cache,
  //     'bl trampoline' is replaced with `bl destination` and the trampoline is not used.
```

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:
```diff
@@ -1113,10 +1113,51 @@ class MacroAssembler: public Assembler {
+  //
+  //     A link-time optimization can be applied to a trampoline call when the generated
+  //     code is moved to the code cache. A trampoline call is replaced with by a direct call.
+  //     The optimization does not remove the trampoline from the stub section.
```
```suggestion
  //     The optimization does not remove the trampoline from the stub section.
  //     This is necessary because the trampoline may well be redirected later when
  //     code is patched, and the new destination may not be reachable by a simple BR
  //     instruction.
```

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:
```diff
@@ -1113,10 +1113,51 @@ class MacroAssembler: public Assembler {
+  //
+  //   - indirect call: move reg, address; blr reg
+  //     This too can reach anywhere in the address space, but it cannot be
+  //     patch while code is running, so it must be used at a safepoint.
```
```suggestion
  //     patched while code is running, so it must be used at a safepoint.
  //     This form of call is most suitable for targets at fixed addresses, whick
  //     will never be patched.
```

theRealAph replied:
Suggest:
```
//
// The patching we do conforms to the "Concurrent modification and
// execution of instructions" section of the Arm Architectural
// Reference Manual, which only allows B, BL, BRK, HVC, ISB, NOP, SMC,
// or SVC instructions to be modified while another thread is
// executing them.
//
// To patch a trampoline call when the BL can't reach, we first modify
// the 64-bit destination address in the trampoline, then modify the
// BL to point to the trampoline, then flush the instruction cache to
// broadcast the change to all executing threads. See
// NativeCall::set_destination_mt_safe for the details.
//
// There is a benign race in that the other thread might observe the
// modified BL before it observes the modified 64-bit destination
// address. That does not matter because the target method has been
// invalidated, so there will be a trap at its start..

```

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:
```diff
@@ -1127,21 +1168,19 @@ class MacroAssembler: public Assembler {
   }
 
-  // Far_call and far_jump generate a call of/jump to the provided address.
+  // Emit a direct call/jump if the entry address is always in range,
```
```suggestion
  // Emit a direct call/jump if the entry address will always in be range,
```

theRealAph commented:
> Hi Andrew,
> 
> The final version is here. Could you please review it?

It might be a good idea to handle the code changes in another patch later. I don't think we should push minor code changes at this point before JDK 19 rampdown, and this is a doc patch.

eastig commented:
> > Hi Andrew,
> > The final version is here. Could you please review it?
> 
> It might be a good idea to handle the code changes in another patch later. I don't think we should push minor code changes at this point before JDK 19 rampdown, and this is a doc patch.

Completely agree. I'll move them into separate PRs:
- Remove `trampoline_call1`
- Remove the `cbuf` argument from far_call/far_jump/trampoline_call.

vnkozlov reviewed (approved):
Looks good and understandable.

eastig commented:
> * Remove `trampoline_call1`

https://bugs.openjdk.java.net/browse/JDK-8287393

> * Remove the `cbuf` argument from far_call/far_jump/trampoline_call.

https://bugs.openjdk.java.net/browse/JDK-8287394

vnkozlov commented:
/sponsor
